### PR TITLE
AXI4 Deinterleaver: update to chisel3 to get inferred reset

### PR DESCRIPTION
**Pull request by mwachs5**
**Related issue**: None

**Type of change**: feature request 

**Impact**: API addition (no impact on existing code)

**Development Phase**:  implementation

**Release Notes**

Support Asynchronous Reset for AXI4 Deinterleaver

### CODE REVIEW

**Comment by mwachs5** [comment]
yeah but I can't actually do that in the GitHub gui :)

**Comment by mwachs5** [comment]
the need for `DontCare`s is killing me. Will push what I have in the "full chisel3 conversion" but getting a bunch of errors about the queues not being fully initialized

File: src/main/scala/amba/axi4/Deinterleaver.scala [hunk 1]
```diff
@@ -46,9 +47,12 @@ class AXI4Deinterleaver(maxReadBytes: Int)(implicit p: Parameters) extends LazyM
           }
         }
 
+        // This seems dubious. Are all these queues not actually used?
+        qs.foreach { _ := DontCare }
```
Comment (mwachs5) [inline src/main/scala/amba/axi4/Deinterleaver.scala:51]: @terpstra why is this OK (if I don't put this i get lots of firrtl errors about things being `VOID`)? At the very least it seems we should be defaulting the readys and valids to 0, not DontCare... down below they only get conditionally set which makes `chisel3` unhappy.

File: src/main/scala/amba/axi4/Deinterleaver.scala [hunk 2]
```diff
@@ -81,13 +85,15 @@ class AXI4Deinterleaver(maxReadBytes: Int)(implicit p: Parameters) extends LazyM
...
+        val deq_bits = WireInit(VecInit(qs.map(_.deq.bits)))
+        in.r.bits  := deq_bits(deq_id)
         (deq_OH.asBools zip qs) foreach { case (s, q) =>
           q.deq.ready := s && in.r.fire()
         }
 
         // Feed response into matching Q
-        out.r.ready := Vec(qs.map(_.enq.ready))(enq_id)
+        val enq_readys = WireInit(VecInit(qs.map(_.enq.ready)))
+        out.r.ready := enq_readys(enq_id)
```
Comment (terpstra) [inline src/main/scala/amba/axi4/Deinterleaver.scala:96]: Why did you need the WireInit?

File: src/main/scala/amba/axi4/Deinterleaver.scala [hunk 3]
```diff
@@ -81,13 +85,15 @@ class AXI4Deinterleaver(maxReadBytes: Int)(implicit p: Parameters) extends LazyM
 
         // Transmit the selected burst to inner
         in.r.valid := locked
-        in.r.bits  := Vec(qs.map(_.deq.bits))(deq_id)
+        val deq_bits = WireInit(VecInit(qs.map(_.deq.bits)))
+        in.r.bits  := deq_bits(deq_id)
```
Comment (terpstra) [inline src/main/scala/amba/axi4/Deinterleaver.scala:89]: Was changing Vec to VecInit not enough?

File: src/main/scala/amba/axi4/Deinterleaver.scala (hunk 2 above)
Comment (mwachs5) [reply src/main/scala/amba/axi4/Deinterleaver.scala:96]: you can's just do Vec(Seq) anymore. I think this is what the cihsel3 cookbook says to do now.

File: src/main/scala/amba/axi4/Deinterleaver.scala (hunk 3 above)
Comment (mwachs5) [reply src/main/scala/amba/axi4/Deinterleaver.scala:89]: TBH i didn't try it. I will try it out. I was going by memory of fixing these things before, but maybe that's only needed for RegInit

File: src/main/scala/amba/axi4/Deinterleaver.scala (hunk 1 above)
Comment (terpstra) [reply src/main/scala/amba/axi4/Deinterleaver.scala:51]: Hmm. I don't understand this, either. The ready and valid seem to be driven unconditionally in the zip+foreach. Is it possible that the zip length is shorter than the 'qs' length?

Comment (terpstra) [reply src/main/scala/amba/axi4/Deinterleaver.scala:51]: Hmm. I don't understand this, either. The ready and valid seem to be driven unconditionally in the zip+foreach. Is it possible that the zip length is shorter than the 'qs' length?

File: src/main/scala/amba/axi4/Deinterleaver.scala (hunk 3 above)
Comment (jackkoenig) [reply src/main/scala/amba/axi4/Deinterleaver.scala:89]: `VecInit` should be sufficient

File: src/main/scala/amba/axi4/Deinterleaver.scala (hunk 2 above)
Comment (mwachs5) [reply src/main/scala/amba/axi4/Deinterleaver.scala:96]: ok you're right `VecInit` alone is enough

File: src/main/scala/amba/axi4/Deinterleaver.scala (hunk 3 above)
Comment (mwachs5) [reply src/main/scala/amba/axi4/Deinterleaver.scala:89]: confirmed

File: src/main/scala/amba/axi4/Deinterleaver.scala (hunk 1 above)
Comment (mwachs5) [reply src/main/scala/amba/axi4/Deinterleaver.scala:51]: i put a require that they be the same length, which did not fail, yet i still get the errors if I dont put in the `DontCare`.

Comment (mwachs5) [reply src/main/scala/amba/axi4/Deinterleaver.scala:51]: @terpstra did something change about `QueueIO`? in the case where you DON't put down a queue for `qs` and just put down a QueueIO, what makes the deq.bits := enq.bits ?

I think that's what its complaining about. I don't know why this would have worked before

Comment (mwachs5) [reply src/main/scala/amba/axi4/Deinterleaver.scala:51]: (talking about this line: https://github.com/chipsalliance/rocket-chip/blob/a344f719bd14db9a9e77b868eb645afba616f232/src/main/scala/amba/axi4/Deinterleaver.scala#L46)

Comment (mwachs5) [reply src/main/scala/amba/axi4/Deinterleaver.scala:51]: ah, the commit that added that line says "axi4: Deinterleaver need not make a Q for an unused AXI id". So presumably these wires are really `DontCare`-able, and we could put an assertion that their enq.fire == false?

Or I will plan to just wire them up so that deq := enq

File: src/main/scala/amba/axi4/Deinterleaver.scala [hunk 4]
```diff
@@ -27,28 +28,39 @@ class AXI4Deinterleaver(maxReadBytes: Int)(implicit p: Parameters) extends LazyM
...
           } else {
-            Wire(new QueueIO(out.r.bits.cloneType, beats))
+            // These are unused IDs and should be never used.
+            // But, to satisfy type checks we must produce a Wire of the
+            // correct type.
+            val q = Wire(new QueueIO(out.r.bits.cloneType, beats))
+            q.suggestName(s"queue_wire_${i}")
+            assert(!q.enq.valid, s"ID ${i} should not be used")
+            // These could just be marked DontCare
+            q.deq :<> q.enq
```
Comment (mwachs5) [inline src/main/scala/amba/axi4/Deinterleaver.scala:55]: @terpstra can you re-review this bit? This is all that is needed to be DontCare, not every queue, but I wired them up to eachother and added an assertion instead

File: src/main/scala/amba/axi4/Deinterleaver.scala [hunk 5]
```diff
@@ -81,14 +93,19 @@ class AXI4Deinterleaver(maxReadBytes: Int)(implicit p: Parameters) extends LazyM
...
-        (deq_OH.asBools zip qs) foreach { case (s, q) =>
+        val deq_bits = VecInit(qs.map(_.deq.bits))
+        in.r.bits  := deq_bits(deq_id)
+        val deq_OH_bools = deq_OH.asBools
+        require(deq_OH_bools.size == qs.size, s"deq_OH.size != qs.size (${deq_OH_bools.size} vs ${qs.size})")
+        (deq_OH_bools zip qs) foreach { case (s, q) =>
           q.deq.ready := s && in.r.fire()
         }
 
+        require(enq_OH_bools.size == qs.size, s"enq_OH.size != qs.size (${enq_OH_bools.size} vs ${qs.size})")
```
Comment (mwachs5) [inline src/main/scala/amba/axi4/Deinterleaver.scala:104]:
```suggestion
        val enq_OH_bools = enq_OH.asBools
        require(enq_OH_bools.size == qs.size, s"enq_OH.size != qs.size (${enq_OH_bools.size} vs ${qs.size})")
```

File: src/main/scala/amba/axi4/Deinterleaver.scala (hunk 4 above)
Comment (terpstra) [reply src/main/scala/amba/axi4/Deinterleaver.scala:55]: Ahh. This makes sense. Thanks for figuring it out! I do not think cross connecting them is wise, however. Use a DontCare instead.

File: src/main/scala/amba/axi4/Deinterleaver.scala [hunk 6]
```diff
@@ -27,28 +28,39 @@ class AXI4Deinterleaver(maxReadBytes: Int)(implicit p: Parameters) extends LazyM
...
+            // These are unused IDs and should be never used.
+            // But, to satisfy type checks we must produce a Wire of the
+            // correct type.
+            val q = Wire(new QueueIO(out.r.bits.cloneType, beats))
+            q.suggestName(s"queue_wire_${i}")
+            assert(!q.enq.valid, s"ID ${i} should not be used")
+            // These could just be marked DontCare
+            q.deq :<> q.enq
+            q.count := 0.U
+            q
```
Comment (mwachs5) [inline src/main/scala/amba/axi4/Deinterleaver.scala:57]:
```suggestion
            q := DontCare
```

File: src/main/scala/amba/axi4/Deinterleaver.scala (hunk 4 above)
Comment (mwachs5) [reply src/main/scala/amba/axi4/Deinterleaver.scala]: done
